### ISSUE TITLE
Implement copy-on-write for all backends

### ISSUE THREAD

**Issue by mratsim** [mratsim/Arraymancer#157]
While I love value semantics and still thinks it is best to avoid this type of questions:
![image](https://user-images.githubusercontent.com/22738317/33070399-2c208da6-ceb8-11e7-83a0-9205040df5ce.png)

I think:
 - easy syntax,
 - behaviour consistency between backends (Cpu, Cuda, OpenCL),
 - efficient memory allocation
 - and performance
cannot be ignored.

Most Arraymancer performance issues come from memory allocation and are due to value semantics. The alternative, using `unsafeView`, `unsafeSlice`, `unsafeSqueeze`, etc is clunky as it litters the codebase with low-level details.

With "copy-on-write", it is possible to be:
  - safe as with value semantics
  - and fast like reference semantics

by only copying when necessary, i.e. when a value is mutable.

Benefits expected:
 - Copies involving non-mutable variables will be shallow by default
 - Slicing of non-mutable variables will be shallow by default
 - Everything involving mutable variables will be deep by default to prevent "unexpected surprises"
 - Keep the behaviour predictable:
      - let assignment from let ==> shallow
      - var assignment from let ==> deep
      - let assignment from var ==> deep
      - There will be no under the hood `if refcount = 1` optimization for var assignment .
         That would make Arraymancer depends on the refcounting GC and it would be much better to use [destructors](https://nim-lang.org/araq/destructors.html) and the future `=move` and `=sink` in #150 for that.
  - Copy-on-write will work well with unsafe proc chaining #151

Limitations:
  - Implicit `result` will still require `unsafeView` and `unsafeUnsqueeze` for shallow-copies. (Obviously slice mutation does not copy but overwrite in-place).
     This is perfectly reasonable in my opinion.

This superseeds #19

**Comment by mratsim** [comment]
Still stuck on assignment overloading: https://github.com/nim-lang/Nim/issues/6348 and https://github.com/nim-lang/Nim/issues/6786

**Comment by mratsim** [comment]
Outcome of my research.

A tentative copy-on-write object to be used in a single-threaded context or a master thread + slave thread managed by OpenMP that only works on already allocated memory:

Forum [thread](https://forum.nim-lang.org/t/3355).

```Nim
{.experimental.}

type
  Storage[T] = ref object
    refcount: int
    data: seq[T]
    # lock / guard ?

type COWobject[T] = object
  metadata: int
  storage: Storage[T]

proc detach[T](c: var COWobject[T]) =
  # Note: this only works without races if only the main thread can access this.
  # Also increment is only done on assignment, slices do not increment.
  if c.storage.refcount == 1:
    return
  
  let old_store = c.storage
  var fresh_store: Storage[T]
  new fresh_store
  
  fresh_store.refcount = 1
  deepCopy(fresh_store.data, old_store.data)
  
  c.storage = fresh_store
  dec old_store.refcount

proc `=`[T](dst: var CowObject[T]; src: CowObject[T]) =
  inc src.storage.refcount
  system.`=`(dst, src)

proc `=destroy`[T](c: CowObject[T]) =
  # Note from Nim manual: destructors are tied to a variable
  # And will not trigger for say slices.
  dec c.storage.refcount

proc toCowObj[T](s: varargs[T]): COWobject[T] {.noInit.} =
  result.metadata = 1337
  new result.storage
  result.storage.data = @s

proc `[]`[T](c: COWobject[T], idx: int): T =
  c.storage.data[idx]

proc `[]`[T](c: var COWobject[T], idx: int): var T =
  detach c
  c.storage.data[idx]

proc `[]=`[T](c: var COWobject[T], idx: int, val: T) =
  detach c
  c.storage.data[idx] = val


proc main() =
  let a = toCowObj(1, 2, 3, 4, 5)
  let b = a
  var c = a
  let d = c
  var e = c
  let f = e
  let g = f
  
  c[1] += 10
  e[2] = 100
  
  
  echo "\n\nMemory location"
  echo "a: [1, 2, 3, 4, 5]: " & $a.repr
  echo "b: [1, 2, 3, 4, 5]: " & $b.repr
  echo "c: [1, 12, 3, 4, 5]: " & $c.repr
  echo "d: [1, 2, 3, 4, 5]: " & $d.repr
  echo "e: [1, 2, 100, 4, 5]: " & $e.repr
  echo "f: [1, 2, 3, 4, 5]: " & $f.repr
  echo "g: [1, 2, 3, 4, 5]: " & $g.repr
  
  echo "\n\n"
  echo "a: [1, 2, 3, 4, 5]: " & $a
  echo "b: [1, 2, 3, 4, 5]: " & $b
  echo "c: [1, 12, 3, 4, 5]: " & $c
  echo "d: [1, 2, 3, 4, 5]: " & $d
  echo "e: [1, 2, 100, 4, 5]: " & $e
  echo "f: [1, 2, 3, 4, 5]: " & $f
  echo "g: [1, 2, 3, 4, 5]: " & $g

main()
```


# Concurrent thread-safe version

Currently there is no plan to make that thread-safe with different Tensors in different thread sharing the same memory. First of all Nim `seq` and GC are thread-local.

In case this is needed is the future, here is some research around copy-on-write, concurrency, lock, lock-free or wait-free data structure


## Nim

Nim offers several intrinsics by default:
- proc atomicInc(memLoc: var int; x: int = 1): int {..}
- proc atomicDec(memLoc: var int; x: int = 1): int {..}
- proc addAndFetch(p: ptr int; val: int): int {..}
- proc atomicInc(memLoc: var int; x: int = 1): int {..}
- proc atomicDec(memLoc: var int; x: int = 1): int {..}
- proc cas[T: bool | int | ptr](p: ptr T; oldValue, newValue: T): bool {..}
- proc cpuRelax() 

[Undocumented intrinsics in atomics.nim](https://github.com/nim-lang/Nim/blob/master/lib/system/atomics.nim#L291)

[The locks module](https://nim-lang.org/docs/locks.html#atomicInc,int,int), with a very nice "withLock" template.


And several [lock examples in the manual](https://nim-lang.org/docs/manual.html#guards-and-locks)

## Warnings - lock-free is not the solution to everything
- [Fear and Loathing in Lock-Free programming](https://medium.com/@tylerneely/fear-and-loathing-in-lock-free-programming-7158b1cdd50c)
- [Single writer principle](https://mechanical-sympathy.blogspot.fr/2011/09/single-writer-principle.html)
Note this very nice table about contention:

Method | Time (ms)
-- | --
One Thread | 300
One Thread with Memory Barrier | 4,700
One Thread with CAS | 5,700
Two Threads with CAS | 18,000
One Thread with Lock | 10,000
Two Threads with Lock | 118,000

- [Intel Threads Building Blocks](https://software.intel.com/en-us/forums/intel-threading-building-blocks/topic/296897) do NOT use lock-free queue because they are slower than locking/spinlocks
- [Another Intel forum topic on non-blocking data structure](https://software.intel.com/en-us/forums/intel-moderncode-for-parallel-architectures/topic/295279)

## Benchmarks
- [Synchrobench](https://github.com/gramoli/synchrobench)
    - arrays, binary trees, hash tables, linked lists, queues and skip lists that are synchronized with copy-on-write, locks, read-modify-write, read-copy-update and transactional memory.

## Reference-counting
- [Refcounting in D](https://bartoszmilewski.com/2009/08/19/the-anatomy-of-reference-counting/): GC and destructors interactions
- Stack overflow discussion of [Boost tutorial implementation](https://stackoverflow.com/questions/31254767/atomic-reference-counting) of atomic reference counting
- [DrDobb's in-depth implementation](http://www.drdobbs.com/atomic-reference-counting-pointers/184401888)
- [Reference counting - Harder than it sounds](https://www.playingwithpointers.com/refcounting-harder-than-it-sounds.html)
- [Lock-free reference counting](https://www.arangodb.com/2015/08/lockfree-protection-of-data-structures-that-are-frequently-read/), hazard pointers and scope guards. Code [here](https://github.com/neunhoef/DataProtector).

## Atomics
- [Herb Sutter on atomics in C++](https://herbsutter.com/2013/02/11/atomic-weapons-the-c-memory-model-and-modern-hardware/)
- [Atomics and memory ordering](https://bartoszmilewski.com/2008/12/01/c-atomics-and-memory-ordering/)

## R/W lock (many readers or one writer)
- [RW spinlock implementation in C++](https://gist.github.com/yizhang82/500da684837161055978011c5850d296#file-rw_spin_lock-h)
- ["lockfree R/W lock"](http://yizhang82.me/lock-free-rw-lock)

## Read-copy-update (RCU)
- [RCU lock](https://github.com/nivekuil/rculock) in Rust using [Crossbeam](https://github.com/crossbeam-rs/crossbeam)
- [In-depth review of RCU](https://kukuruku.co/post/lock-free-data-structures-the-inside-rcu/)

## Copy on write
- [C++ example of copy-on-write strings](https://gist.github.com/Manu343726/02287de75bb24f2cef00)
- [C++14 "Copy on write pointer done right"](https://github.com/etam/cowptr)
- [Partial copy-on-write Deferred Reclamation (PDR)](https://groups.google.com/forum/#!topic/comp.programming.threads/fSAucmQkwL8)
- [Limit of copy-on-write in Ruby](https://brandur.org/ruby-memory)
- To be checked: R implementation

## Software Transactional Memory
- [Wikipedia](https://en.wikipedia.org/wiki/Software_transactional_memory)
- [StackOverflow discussion](https://stackoverflow.com/questions/1627600/how-do-you-implement-software-transactional-memory)
 - [Beyond locks: Software Transactional Memory](https://bartoszmilewski.com/2010/09/11/beyond-locks-software-transactional-memory/)

**Comment by mratsim** [comment]
WIP in https://github.com/mratsim/Arraymancer/pull/159

## Further research
- Herb Sutter on [CoW pessimization](http://www.cpptips.com/mt_ref_count) and [in-depth article](http://www.gotw.ca/publications/optimizations.htm)
- Using [deep-copy in Perl to avoid CoW performance issues](https://retout.co.uk/blog/2012/12/21/)
- [Instagram engineers: Is Python refcounting forcing copy on read](https://engineering.instagram.com/dismissing-python-garbage-collection-at-instagram-4dca40b29172) (since read takes a reference in Python)

**Comment by mratsim** [comment]
After tinkering for a few hours, I have definite proof that copy-on-write does not work cleanly for a neural net library so I will use reference semantics by default and require cloning for deep copies.

- Increasing and decreasing the refcounter in tight loops seems quite costly, I didn’t benchmark properly on my full bench suite, but my incRef/decRef proc were taking 1-3% on a `test_nnp_loss`  (sigmoid_cross_entropy and softmax_cross_entropy)

- Bis. Potential idea, instead of having a counter: just have a boolean “shared”, when doing assignment, changed shared to true, when doing `=sink`do not change it. —> when “shared is true, mutation copies”. This avoids keeping track of the refcounter.

- Except that at the neural network level, Tensors will be referred to by a graph and they will always be “shared”, defeating the whole purpose of copy-on-write

Last things
- implicit sharing means hard to know if Arraymancer would implicitly copy or implicitly share, meaning users will have to think if this is important.
- to provide "unsafe" semantics with COW tensors, the syntax is system.`=`(dest, src) which is also unwieldly (with deep copy, just use unsafeView, with ref semantics, just use clone)

**Comment by mratsim** [comment]
Closed by #160
